search/searcher/disjunction.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package searcher
+
+import (
+	"github.com/m3db/m3ninx/postings"
+	"github.com/m3db/m3ninx/search"
+)
+
+type disjunctionSearcher struct {
+	searchers search.Searchers
+	len       int
+
+	idx  int
+	curr postings.List
+
+	closed bool
+	err    error
+}
+
+// NewDisjunctionSearcher returns a new Searcher which matches documents which are matched
+// by any of the given Searchers. It is not safe for concurrent access.
+func NewDisjunctionSearcher(ss search.Searchers) (search.Searcher, error) {
```

Comment:
not sure I understand the intent of this code. 

Here's my take: for input it takes a series of searchers, which are effectively iterators over postings lists. Then you take the union of these indexwise. 

i.e. 
```
// Given input: 
  S1: [P_11, P_12, ...]
  S2: [P_21, P_22, ...]

// The output this gives back is:
  [P_11 ∪ P_21, P_12 ∪ P_22, ... ]
```

Is that right?

Reply:
That's correct, it's meant to be segment-wise though. Doing it across indices would require a little more care since we lose postings list uniqueness.

Reply:
Fair. Could you put a note down in the interface explaining this semantic in more detail

Reply:
Yep, added a note to the documentation on the interface.